### PR TITLE
Parallel write storage

### PR DESCRIPTION
**Pull request by shargon**
Possible fix for https://github.com/neo-project/neo/issues/694

### CODE REVIEW

File: neo/IO/Caching/DataCache.cs [hunk 1]
```diff
@@ -62,7 +63,8 @@ public void Add(TKey key, TValue value)
 
         public void Commit()
         {
-            foreach (Trackable trackable in GetChangeSet())
+            Parallel.ForEach(GetChangeSet(), (trackable) =>
```
Comment (erikzhang) [inline neo/IO/Caching/DataCache.cs]: I don't think it works. We use batch in inner db.

**Comment by erikzhang** [comment]
What does this PR for?

**Comment by shargon** [comment]
Is for try to prevent a denial of service when we upgrade a big contract, i think that it could be solved with tasks

**Comment by igormcoelho** [comment]
In fact Shargon, in the case of massive migration, it's better to solve this with MPT storages. A single storage key changes the whole storage location @rodoufu 
The proposal by Erik is in fact much deeper, it means that you could invoke some contract, and it "automatically" invoke another one (migrated from the first one).
The point there is:
- Do we want/need this feature?
- How to implement it?

To implement it, my opinion is to have two different types of calls, one accepting direction, and another that won't accept. There are situations where you must be sure that scripthash matches exactly.

One example:
- Withdraw funds from contractA on verification. This will invoke contractA... but wait, contractA now is contractB (after migration), but which funds will be taken back, contractA, or contractB? What happens if I put funds on contractA, after migration to contractB? Should it migrate from contractA to B? Will it allow another future deploy on contractA after migration to contractB?

In my opinion, the PR should address all (or most of) these questions, in order to solve the contract redirection proposal by Erik.

File: neo/IO/Caching/DataCache.cs (hunk 1 above)
Comment (shargon) [reply neo/IO/Caching/DataCache.cs]: Maybe i can remove this, and only maintain the `Contract Update` changes?

**Comment by erikzhang** [comment]
> Is for try to prevent a denial of service when we upgrade a big contract, i think that it could be solved with tasks

I think it makes things worse. We need to consider #694.

**Comment by shargon** [comment]
I have an idea for https://github.com/neo-project/neo/issues/694

**Pull request closed without merging**